src/engine/cachingreader.cpp

```diff
@@ -200,43 +200,39 @@ CachingReaderChunkForOwner* CachingReader::lookupChunkAndFreshen(SINT chunkIndex
     return pChunk;
 }
 
+// Invoked from the UI thread!!
 void CachingReader::newTrack(TrackPointer pTrack) {
-    // Feed the track to the worker as soon as possible
-    // to get ready while the reader switches its internal
-    // state. There are no race conditions, because the
-    // reader polls the worker.
-    m_worker.newTrack(pTrack);
-    m_worker.workReady();
-    // Don't accept any new read requests until the current
-    // track has been unloaded and the new track has been
-    // loaded.
-    m_state = State::TrackLoading;
-    // Free all chunks with sample data from the current track.
-    freeAllChunks();
+    auto newState = pTrack ? STATE_TRACK_LOADING : STATE_TRACK_UNLOADING;
+    auto oldState = m_state.fetchAndStoreAcquire(newState);
+    VERIFY_OR_DEBUG_ASSERT(
+            oldState != STATE_TRACK_LOADING ||
+            newState != STATE_TRACK_LOADING) {
+        kLogger.critical()
+                << "Cannot load new track while loading a track";
+        return;
+    }
+    m_worker.newTrack(std::move(pTrack));
```

Comment:
The final line `m_worker.workReady();` got lost along the way

Reply:
from your template :-)